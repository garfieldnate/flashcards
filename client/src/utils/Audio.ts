import { Audio } from 'expo-av';
import { Sound } from 'expo-av/build/Audio';
import { PlaybackSource } from 'expo-av/build/AV';

export const loadAudio = async (src: PlaybackSource): Promise<Sound> => {
  const sound = new Audio.Sound();
  await sound.loadAsync(src);
  return sound;
};
