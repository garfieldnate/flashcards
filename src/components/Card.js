// Visual component for a single flashcard
import PropTypes from 'prop-types';
import { Audio } from 'expo';
import React, { Component } from 'react';
import { Platform } from 'react-native';
import {
  StyleSheet,
  Text,
  TouchableOpacity,
  View
} from 'react-native';
import cardLayout from './CardLayout.js';

import CardFlip from 'react-native-card-flip';

export default class Card extends Component<Props> {
  static propTypes = {
    cardID: PropTypes.number.isRequired,
    front: PropTypes.string.isRequired,
    back: PropTypes.string.isRequired,
    exampleForeignLang: PropTypes.string,
    exampleUserLang: PropTypes.string,
    // result of asset require() is a number
    foreignHeadwordAudio: PropTypes.number
  };

  constructor (props) {
    super(props);
    if(this.props.foreignHeadwordAudio) {
      this.recording = new Audio.Sound();
      this.recording.loadAsync(this.props.foreignHeadwordAudio).
        then(() => {
          this.recordingReady = true;
          console.log(`loaded audio for ${this.props.back}`)
        }).
        catch((error) => {
          console.log(`Couldn't load audio for ${this.props.back}`);
          console.log(error);
        });
    }
  }

  render () {
    return (
        <CardFlip style={styles.cardContainer} ref={(card) => this.card = card} >
          <View style={[styles.card, styles.cardFront]}>
            {this.renderTopSection(this.props.front)}
            <View style={styles.cardBottomSection} />
          </View>
          <View style={[styles.card, styles.cardBack]} >
            {this.renderTopSection(this.props.back)}
            <View style={styles.cardBottomSection}>
              {this.renderExample()}
            </View>
          </View>
        </CardFlip>
    );
  }

  renderTopSection = (headwordText) => {
    return (
      <View style={{flex: 2, justifyContent: 'flex-end'}}>
        <View style={styles.textContainer} >
          <Text adjustsFontSizeToFit numberOfLines={countLines(headwordText)} style={styles.headword}>{headwordText}</Text>
        </View>
      </View>);
  }

  // Renders the foreign language example, the user's language example,
  // and a separating line between them. Adjusts for missing sentence
  // data by omitting elements as needed.
  renderExample = () => {
    var foreignExample, separator, userLangExample;
    if (!this.props.exampleForeignLang && !this.props.exampleUserLang) {
      return null;
    }

    if(this.props.exampleForeignLang) {
      foreignExample = <Text style={styles.exampleText}>{this.props.exampleForeignLang}</Text>;
    }
    if(this.props.exampleUserLang) {
      userLangExample = <Text style={styles.exampleText}>{this.props.exampleUserLang}</Text>;
    }
    if(foreignExample && userLangExample) {
      separator = <Text style={styles.exampleText}>{'─────'}</Text>;
    }
    return (
      <View style={[styles.textContainer, styles.exampleContainer]}>
        {foreignExample}
        {separator}
        {userLangExample}
      </View>);
  }

  flip = () => {
    if (this.card.state.side === 0) {
      if(this.recording) {
        console.log(`playing recording ${this.props.back}`);
        this.recording.replayAsync().
          catch((error) => {
            console.log("Error while playing mp3");
            console.log(error);
          });
      }
    }
    this.card.flip();
  }
}

const countLines = (text) => {
  return (text.match(/\r?\n/g)||[]).length + 1;
}

const styles = StyleSheet.create({
  cardContainer:{
    flex: 1
  },
  card: {
    shadowColor: 'rgba(0,0,0,0.5)',
    shadowOffset: {
      width: 0,
      height: 1
    },
    shadowOpacity:0.5,
    overflow: 'hidden',
    ...cardLayout
  },
  cardFront: {
    backgroundColor: '#91CB3E',
  },
  cardBack: {
    backgroundColor: '#84B939',
  },
  textContainer: {
    backgroundColor: 'rgba(0,0,0,0.4)',
  },
  cardBottomSection: {
    flex: 3,
    justifyContent: 'space-around'
  },
  headword: {
    textAlign: 'center',
    fontSize: 55,
    color: '#ffffff',
  },
  exampleContainer: {
    padding: '2%',
    // overflow: 'hidden',
    flexShrink: 1,
  },
  exampleText: {
    textAlign: 'center',
    fontSize: 18,
    color: '#FFFFFF',
    fontWeight: '500',
  }
});
