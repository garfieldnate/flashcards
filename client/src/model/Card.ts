import { Sound } from 'expo-av/build/Audio';
import { ImageSourcePropType } from 'react-native';

export interface ICard {
  getId(): string;
  getHeadwordForeignLang(): string;
  getCategory(): string;
  getExampleForeignLang(): string | undefined;
  getExampleUserLang(userLang: string): string | undefined;
  getHeadwordUserLang(userLang: string): string;
  getForeignHeadwordAudio(): Promise<Sound | undefined>;
  getImage(): Promise<ImageSourcePropType | undefined>;
}
