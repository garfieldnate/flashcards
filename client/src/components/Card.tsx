// Visual component for a single flashcard
import { Audio } from 'expo-av';
import React, { Component } from 'react';
import { StyleSheet, Text, View } from 'react-native';
import { ICard as CardData } from '../model/Card';
import cardLayout from './CardLayout';

import { Sound } from 'expo-av/build/Audio';
import CardFlip from 'react-native-card-flip';
import DeleteButton from './DeleteButton';
// import createCardCollection from '../db/CardsDB';

interface IProps {
  cardData: CardData;
  firstTimeSeen?: boolean;
  onDelete: (cardData: CardData) => void;
}

export default class Card extends Component<IProps> {
  private recording?: Sound;
  private recordingReady: boolean = false;
  private cardRef = React.createRef<CardFlip>();

  constructor(props: IProps) {
    super(props);
    // console.log(`cardData is ${this.props.cardData.getHeadwordUserLang('')}`);
    this.props.cardData
      .getForeignHeadwordAudio()
      .then((result) => {
        result.ifPresentOrElse(
          (sound) => {
            this.recording = sound;
            this.recordingReady = true;
            console.log(
              `loaded audio for ${this.props.cardData.getHeadwordForeignLang()}`
            );
          },
          () => {
            'Sound was not present';
          }
        );
      })
      .catch((error) => {
        console.log(
          `Couldn't load audio for ${this.props.cardData.getHeadwordForeignLang()}`
        );
        console.log(error);
      });
  }

  public render() {
    const onDelete = () => this.props.onDelete(this.props.cardData);
    return (
      <CardFlip style={styles.cardContainer} ref={this.cardRef}>
        <View style={[styles.card, styles.cardFront]}>
          <View style={styles.deleteButton}>
            <DeleteButton onPress={onDelete} />
          </View>
          {this.renderTopSection(
            this.props.cardData.getHeadwordUserLang('TODO: user lang here')
          )}
          <View style={styles.cardBottomSection} />
        </View>
        <View style={[styles.card, styles.cardBack]}>
          {this.renderTopSection(this.props.cardData.getHeadwordForeignLang())}
          <View style={styles.cardBottomSection}>{this.renderExample()}</View>
        </View>
      </CardFlip>
    );
  }

  public renderTopSection = (headwordText: string) => {
    return (
      <View style={{ flex: 2, justifyContent: 'flex-end' }}>
        <View style={styles.textContainer}>
          <Text
            adjustsFontSizeToFit
            numberOfLines={countLines(headwordText)}
            style={styles.headword}
          >
            {headwordText}
          </Text>
        </View>
      </View>
    );
  };

  // Renders the foreign language example, the user's language example,
  // and a separating line between them. Adjusts for missing sentence
  // data by omitting elements as needed.
  public renderExample = () => {
    let foreignExample: JSX.Element | undefined;
    let separator: JSX.Element | undefined;
    let userLangExample: JSX.Element | undefined;
    if (
      this.props.cardData.getExampleForeignLang().isEmpty() &&
      this.props.cardData.getExampleUserLang('TODO: user lang here').isEmpty()
    ) {
      return null;
    }

    if (this.props.cardData.getExampleForeignLang().isPresent()) {
      foreignExample = (
        <Text style={styles.exampleText}>
          {this.props.cardData.getExampleForeignLang().get()}
        </Text>
      );
    }
    if (
      this.props.cardData.getExampleUserLang('TODO: user lang here').isPresent()
    ) {
      userLangExample = (
        <Text style={styles.exampleText}>
          {this.props.cardData.getExampleUserLang('TODO: user lang here').get()}
        </Text>
      );
    }
    if (foreignExample && userLangExample) {
      separator = <Text style={styles.exampleText}>{'─────'}</Text>;
    }
    return (
      <View style={[styles.textContainer, styles.exampleContainer]}>
        {foreignExample}
        {separator}
        {userLangExample}
      </View>
    );
  };

  public flip = () => {
    console.log('flipping!');
    const card = this.cardRef.current!;
    if (card.state.side === 0) {
      if (this.recording && this.recordingReady) {
        console.log(
          `playing recording ${this.props.cardData.getHeadwordForeignLang()}`
        );
        this.recording.replayAsync().catch((error) => {
          console.log('Error while playing mp3');
          console.log(error);
        });
      } else {
        console.log(
          `Didn't play because ${this.recording} and ${this.recordingReady}`
        );
      }
    } else {
      console.log(`Didn't play because side is ${card.state.side}`);
    }
    card.flip();
  };
}

const countLines = (text: string) => {
  return (text.match(/\r?\n/g) || []).length + 1;
};

const styles = StyleSheet.create({
  card: {
    overflow: 'hidden',
    shadowColor: 'rgba(0,0,0,0.5)',
    shadowOffset: {
      height: 1,
      width: 0,
    },
    shadowOpacity: 0.5,
    ...cardLayout,
  },
  cardBack: {
    backgroundColor: '#84B939',
  },
  cardBottomSection: {
    flex: 3,
    justifyContent: 'space-around',
  },
  cardContainer: {
    flex: 1,
  },
  cardFront: {
    backgroundColor: '#91CB3E',
  },
  deleteButton: {
    // flex: 1,
    left: 5,
    position: 'absolute',
    top: 5,
    zIndex: 99999,
  },
  exampleContainer: {
    flexShrink: 1,
    padding: '2%',
    // overflow: 'hidden',
  },
  exampleText: {
    color: '#FFFFFF',
    fontSize: 18,
    fontWeight: '500',
    textAlign: 'center',
  },
  headword: {
    color: '#ffffff',
    fontSize: 55,
    textAlign: 'center',
  },
  textContainer: {
    backgroundColor: 'rgba(0,0,0,0.4)',
  },
});
