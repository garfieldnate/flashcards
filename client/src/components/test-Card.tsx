// import React from 'react';
import { Sound } from 'expo-av/build/Audio';
import * as React from 'react';
import { ImageURISource } from 'react-native';
import * as renderer from 'react-test-renderer';
import { ICard } from '../model/Card';
import Card from './Card';

class DummyCard implements ICard {
  public getId(): string {
    return '1';
  }
  public getHeadwordForeignLang(): string {
    return 'ハロー';
  }
  public getCategory(): string {
    return 'foo';
  }
  public getExampleForeignLang(): string | undefined {
    return '例文';
  }
  public getExampleUserLang(userLang: string): string | undefined {
    return 'example';
  }
  public getHeadwordUserLang(userLang: string): string {
    return 'hello';
  }
  public getForeignHeadwordAudio(): Promise<Sound | undefined> {
    return Promise.resolve(undefined);
  }
  public getImage(): Promise<
    number | ImageURISource | ImageURISource[] | undefined
  > {
    return Promise.resolve(undefined);
  }
}

it('renders a card', () => {
  const tree = renderer
    .create(<Card cardData={new DummyCard()} onDelete={() => {}} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
