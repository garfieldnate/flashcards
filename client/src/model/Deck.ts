import { Card } from './Card';
export interface IDeck {
  ID: string;
  name: string;
  thumbnail: string;
  cardsDue: number;
  cards: Card[];
}
