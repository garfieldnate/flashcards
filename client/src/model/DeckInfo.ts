import { ImageSourcePropType } from 'react-native';
import { Optional } from 'typescript-optional';
import { CardId, ICard } from './Card';

export interface IDeckInfo {
  getId(): string;
  getName(): string;
  getThumbnail(): ImageSourcePropType;
  getPresentationOrder(): CardId[];
  /**
   *
   * @param id of card to retrieve
   * @returns the card if it is built-in, or undefined if it is not
   */
  getBuiltin(id: CardId): Optional<ICard>;
}
