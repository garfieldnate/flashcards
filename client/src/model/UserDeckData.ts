import { Card } from './Card';

export type UserDeckData = {
  prefs: {
    numNewCardsPerDay: number;
  };
  studyState: {
    cardData: {
      getCardsDueBetween: (
        start: number,
        end: number
      ) => {
        cardsDue: Card[];
        nextDue: number;
      };
    };
    lastStudied: string;
    numAddedToday: number;
  };
};
