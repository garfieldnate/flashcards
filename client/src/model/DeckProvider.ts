import { IDeckInfo } from './Deck';
export default interface IDeckProvider {
  // TODO: return the object here instead of an array
  getAvailableDecks(): IDeckInfo[];
  getDeck(deckID: string): IDeckInfo;
}
