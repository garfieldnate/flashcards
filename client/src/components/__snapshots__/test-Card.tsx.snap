// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a card 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "transform": Array [
        Object {
          "scale": 1,
        },
      ],
    }
  }
>
  <View
    style={
      Object {
        "bottom": 0,
        "flex": 1,
        "left": 0,
        "opacity": 100,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "transform": Array [
          Object {
            "perspective": 800,
          },
          Object {
            "rotateY": "0deg",
          },
        ],
        "zIndex": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "borderColor": "white",
            "borderWidth": 10,
            "height": "85%",
            "overflow": "hidden",
            "shadowColor": "rgba(0,0,0,0.5)",
            "shadowOffset": Object {
              "height": 1,
              "width": 0,
            },
            "shadowOpacity": 0.5,
            "width": "100%",
          },
          Object {
            "backgroundColor": "#91CB3E",
          },
        ]
      }
    >
      <View
        style={
          Object {
            "left": 5,
            "position": "absolute",
            "top": 5,
            "zIndex": 99999,
          }
        }
      >
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "alignSelf": "center",
              "backgroundColor": "#D95C2E",
              "borderColor": "#FAECE7",
              "borderRadius": 99999,
              "borderWidth": 4,
              "height": 60,
              "justifyContent": "center",
              "opacity": 1,
              "textAlign": "center",
              "width": 60,
            }
          }
        >
          <Text />
        </View>
      </View>
      <View
        style={
          Object {
            "flex": 2,
            "justifyContent": "flex-end",
          }
        }
      >
        <View
          style={
            Object {
              "backgroundColor": "rgba(0,0,0,0.4)",
            }
          }
        >
          <Text
            adjustsFontSizeToFit={true}
            numberOfLines={1}
            style={
              Object {
                "color": "#ffffff",
                "fontSize": 55,
                "textAlign": "center",
              }
            }
          >
            hello
          </Text>
        </View>
      </View>
      <View
        style={
          Object {
            "flex": 3,
            "justifyContent": "space-around",
          }
        }
      />
    </View>
  </View>
  <View
    style={
      Object {
        "bottom": 0,
        "flex": 1,
        "left": 0,
        "opacity": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "transform": Array [
          Object {
            "perspective": -800,
          },
          Object {
            "rotateY": "180deg",
          },
        ],
        "zIndex": 0,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "borderColor": "white",
            "borderWidth": 10,
            "height": "85%",
            "overflow": "hidden",
            "shadowColor": "rgba(0,0,0,0.5)",
            "shadowOffset": Object {
              "height": 1,
              "width": 0,
            },
            "shadowOpacity": 0.5,
            "width": "100%",
          },
          Object {
            "backgroundColor": "#84B939",
          },
        ]
      }
    >
      <View
        style={
          Object {
            "flex": 2,
            "justifyContent": "flex-end",
          }
        }
      >
        <View
          style={
            Object {
              "backgroundColor": "rgba(0,0,0,0.4)",
            }
          }
        >
          <Text
            adjustsFontSizeToFit={true}
            numberOfLines={1}
            style={
              Object {
                "color": "#ffffff",
                "fontSize": 55,
                "textAlign": "center",
              }
            }
          >
            ハロー
          </Text>
        </View>
      </View>
      <View
        style={
          Object {
            "flex": 3,
            "justifyContent": "space-around",
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "backgroundColor": "rgba(0,0,0,0.4)",
              },
              Object {
                "flexShrink": 1,
                "padding": "2%",
              },
            ]
          }
        >
          <Text
            style={
              Object {
                "color": "#FFFFFF",
                "fontSize": 18,
                "fontWeight": "500",
                "textAlign": "center",
              }
            }
          >
            例文
          </Text>
          <Text
            style={
              Object {
                "color": "#FFFFFF",
                "fontSize": 18,
                "fontWeight": "500",
                "textAlign": "center",
              }
            }
          >
            ─────
          </Text>
          <Text
            style={
              Object {
                "color": "#FFFFFF",
                "fontSize": 18,
                "fontWeight": "500",
                "textAlign": "center",
              }
            }
          >
            example
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
